File: Sources/hearsay/Engine.swift
import Foundation
import Transcription

enum PrivacyClass: Equatable {
    case onDevice
    case cloud
}

/// The general LLMs hearsay will transcribe with via OpenRouter. One at a time: the dedicated ASR
/// engines are the product, this is the comparison point. Adding one is one case + one row.
enum OpenRouterModel: String, CaseIterable, Equatable {
    case geminiFlash = "google/gemini-3.7-flash"

    var pricePer100kWords: String {
        switch self {
        case .geminiFlash: return "~$1.45 per 100k words"
        }
    }

    var label: String {
        switch self {
        case .geminiFlash: return "Gemini 3.7 Flash"
        }
    }
}

/// The engine concept: who turns audio into text, at what cost and privacy.
/// One type owns identity (wire key), display, availability, and construction. A new engine is one
/// new case here — plus, if it needs an on-device model, a provisioning path in the Coordinator.
enum Engine: Equatable {
    case appleLocal
    case openRouter(OpenRouterModel)
    case elevenLabsScribe
    case geminiTranscribeLive

    static var all: [Engine] {
        [.appleLocal, .elevenLabsScribe, .geminiTranscribeLive] + OpenRouterModel.allCases.map { .openRouter($0) }
    }

    /// Wire key: persisted in Settings and stamped on BakeoffRecords. `init(wireKey:)` is its exact inverse.
    var wireKey: String {
        switch self {
        case .appleLocal: return "apple-local"
        case .openRouter(let model): return model.rawValue
        case .elevenLabsScribe: return "elevenlabs/\(ElevenLabsTranscriber.modelID)"
        case .geminiTranscribeLive: return "google/\(GeminiLiveTranscriber.modelID)"
        }
    }

    init?(wireKey: String) {
        if let match = Engine.all.first(where: { $0.wireKey == wireKey }) {
            self = match
        } else {
            return nil
        }
    }

    var label: String {
        switch self {
        case .appleLocal: return "Apple on-device ($0)"
        case .openRouter(let model): return "Google · \(model.label) (via OpenRouter)"
        case .elevenLabsScribe: return "ElevenLabs · Scribe v2"
        case .geminiTranscribeLive: return "Google · Gemini 3.5 Transcribe (live)"
        }
    }

    var detail: String {
        switch self {
        case .appleLocal: return "SpeechAnalyzer on the Neural Engine, works offline · $0"
        case .openRouter(let model): return "General LLM listening to the audio, via OpenRouter · \(model.pricePer100kWords)"
        case .elevenLabsScribe: return "ElevenLabs Scribe v2 cloud, dedicated ASR, 90+ languages, mixes them mid-sentence · ~$2.45 per 100k words"
        case .geminiTranscribeLive: return "Google cloud, streaming ASR with live partials, 85+ languages, mixes them mid-sentence · ~$6 per 100k words, free tier in preview"
        }
    }

    /// Two or three words: chips, pills and leaderboard rows.
    var shortLabel: String {
        switch self {
        case .appleLocal: return "Apple"
        case .openRouter(let model): return model.label
        case .elevenLabsScribe: return "Scribe v2"
        case .geminiTranscribeLive: return "Gemini Live"
        }
    }

    /// Streams text while you speak. In a race the pill shows the first such engine's partials.
    var deliversPartials: Bool {
        switch self {
        case .appleLocal, .geminiTranscribeLive: return true
        case .openRouter, .elevenLabsScribe: return false
        }
    }

    /// Only the Apple engine needs a locale; the cloud engines detect language themselves.
    /// Exhaustive on purpose: a new engine must decide this explicitly.
    var needsLocale: Bool {
        switch self {
        case .appleLocal: return true
        case .openRouter, .elevenLabsScribe, .geminiTranscribeLive: return false
        }
    }

    var requiredKey: String? {
        switch self {
        case .appleLocal: return nil
        case .openRouter: return "OPENROUTER_API_KEY"
        case .elevenLabsScribe: return "ELEVEN_LABS_API_KEY"
        case .geminiTranscribeLive: return "GEMINI_API_KEY"
        }
    }

    /// Cheap: KeyStore caches the key file until it changes.
    var isAvailable: Bool {
        guard let requiredKey else { return true }
        return KeyStore.value(requiredKey) != nil
    }

    var privacyClass: PrivacyClass {
        switch self {
        case .appleLocal: return .onDevice
        case .openRouter, .elevenLabsScribe, .geminiTranscribeLive: return .cloud
        }
    }

    func makeTranscriber(locale: Locale) -> (any Transcriber)? {
        switch self {
        case .appleLocal: return SpeechAnalyzerTranscriber(locale: locale)
        case .openRouter(let model): return OpenRouterTranscriber(model: model.rawValue)
        case .elevenLabsScribe: return ElevenLabsTranscriber()
        case .geminiTranscribeLive: return GeminiLiveTranscriber()
        }
    }
}

File: crossplatform/crates/core/src/engine.rs
use crate::keystore::KeyStore;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum PrivacyClass {
    OnDevice,
    Cloud,
}

/// whisper.cpp models hearsay knows how to fetch and label. Adding one is one variant + one row.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum WhisperModel {
    BaseEn,
    LargeV3Turbo,
}

impl WhisperModel {
    pub const ALL: [WhisperModel; 2] = [WhisperModel::BaseEn, WhisperModel::LargeV3Turbo];

    pub fn id(self) -> &'static str {
        match self {
            WhisperModel::BaseEn => "base.en",
            WhisperModel::LargeV3Turbo => "large-v3-turbo",
        }
    }

    pub fn file_name(self) -> String {
        format!("ggml-{}.bin", self.id())
    }

    pub fn download_url(self) -> String {
        format!("https://huggingface.co/ggerganov/whisper.cpp/resolve/main/{}", self.file_name())
    }

    pub fn blurb(self) -> &'static str {
        match self {
            WhisperModel::BaseEn => "English only, ~150 MB, fast",
            WhisperModel::LargeV3Turbo => "99 languages with auto-detect, ~1.6 GB, best local accuracy",
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum OpenRouterModel {
    GeminiFlashLite,
    GeminiFlash,
}

impl OpenRouterModel {
    pub const ALL: [OpenRouterModel; 2] = [OpenRouterModel::GeminiFlashLite, OpenRouterModel::GeminiFlash];

    pub fn id(self) -> &'static str {
        match self {
            OpenRouterModel::GeminiFlashLite => "google/gemini-2.5-flash-lite",
            OpenRouterModel::GeminiFlash => "google/gemini-2.5-flash",
        }
    }

    pub fn price_per_100k_words(self) -> &'static str {
        match self {
            OpenRouterModel::GeminiFlashLite => "~$0.50 per 100k words",
            OpenRouterModel::GeminiFlash => "~$1.85 per 100k words",
        }
    }
}

/// The engine concept: who turns audio into text, at what cost and privacy. One type owns identity
/// (wire key), display, availability and privacy. Wire keys are shared with the macOS app so bake-off
/// records are comparable across platforms.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Engine {
    Whisper(WhisperModel),
    OpenRouter(OpenRouterModel),
    ElevenLabsScribe,
}

pub const ELEVENLABS_MODEL_ID: &str = "scribe_v2";

impl Engine {
    pub fn all() -> Vec<Engine> {
        let mut all: Vec<Engine> = WhisperModel::ALL.iter().map(|m| Engine::Whisper(*m)).collect();
        all.push(Engine::ElevenLabsScribe);
        all.extend(OpenRouterModel::ALL.iter().map(|m| Engine::OpenRouter(*m)));
        all
    }

    /// Persisted in settings and stamped on bake-off records. `parse` is its exact inverse.
    pub fn wire_key(&self) -> String {
        match self {
            Engine::Whisper(m) => format!("whisper/{}", m.id()),
            Engine::OpenRouter(m) => m.id().to_string(),
            Engine::ElevenLabsScribe => format!("elevenlabs/{ELEVENLABS_MODEL_ID}"),
        }
    }

    pub fn parse(wire_key: &str) -> Option<Engine> {
        Engine::all().into_iter().find(|e| e.wire_key() == wire_key)
    }

    pub fn label(&self) -> String {
        match self {
            Engine::Whisper(m) => format!("Whisper · {} (local, $0)", m.id()),
            Engine::OpenRouter(m) => format!("OpenRouter · {}", m.id()),
            Engine::ElevenLabsScribe => "ElevenLabs · Scribe".to_string(),
        }
    }

    pub fn detail(&self) -> String {
        match self {
            Engine::Whisper(m) => format!("whisper.cpp on this machine, works offline · {}", m.blurb()),
            Engine::OpenRouter(m) => format!("Google cloud via OpenRouter · {}", m.price_per_100k_words()),
            Engine::ElevenLabsScribe => "ElevenLabs Scribe v2 cloud, dedicated ASR, 90+ languages, mixes them mid-sentence · ~$2.45 per 100k words".to_string(),
        }
    }

    pub fn required_key(&self) -> Option<&'static str> {
        match self {
            Engine::Whisper(_) => None,
            Engine::OpenRouter(_) => Some("OPENROUTER_API_KEY"),
            Engine::ElevenLabsScribe => Some("ELEVEN_LABS_API_KEY"),
        }
    }

    pub fn is_available(&self, keys: &KeyStore) -> bool {
        match self.required_key() {
            None => true,
            Some(name) => keys.value(name).is_some(),
        }
    }

    pub fn privacy_class(&self) -> PrivacyClass {
        match self {
            Engine::Whisper(_) => PrivacyClass::OnDevice,
            Engine::OpenRouter(_) | Engine::ElevenLabsScribe => PrivacyClass::Cloud,
        }
    }

    pub fn default_engine() -> Engine {
        Engine::Whisper(WhisperModel::BaseEn)
    }
}

#[cfg(test)]
mod tests {
    use super::Engine;

    #[test]
    fn wire_key_round_trips_for_every_engine() {
        for engine in Engine::all() {
            assert_eq!(Engine::parse(&engine.wire_key()), Some(engine));
        }
        assert_eq!(Engine::parse("apple-local"), None);
        assert_eq!(Engine::parse("evil-model"), None);
    }
}
